### PR TITLE
API for an async implementation of task-maker-rust. 

### PR DESCRIPTION
**Pull request by veluca93**
The async implementation should be simpler to reason about, allow to
more easily implement some missing features (i.e. streaming of stdout)
and ideally resolve issues due to mixing threads and fork.

### CODE REVIEW

**Comment by veluca93** [comment]
@robryk for discussion

File: task-maker-async/src/store.rs
Comment (robryk) [inline task-maker-async/src/store.rs:12]: Nit: have the second one include everything, so that there's no temptation to extract something out of comparing that alone.

File: task-maker-async/src/store.rs
Comment (robryk) [inline task-maker-async/src/store.rs:56]: I think this is not the same kind of lease that append_input_chunk is talking about. If so, please name them in a distinguishuable way.

Comment (robryk) [reply task-maker-async/src/store.rs:56]: The (a) lack of distinction between creation and refresh and (b) lack of lease identity make it easier for bugs where leases aren't refreshed to hide if someone else is usually keeping a lease on the same file at the same time (or just if the GC doesn't manage to get rid of the file in time).

Maybe have a handle for a lease and refresh/create one separately?

File: task-maker-async/src/store.rs
Comment (robryk) [inline task-maker-async/src/store.rs:103]: Why? What should we do if the previous writer for this computation died?

File: task-maker-async/src/store.rs
Comment (robryk) [inline task-maker-async/src/store.rs:80]: What if the file is not finalized? If we can open nonfinalized files, then what if the file is never finalized (e.g. because its writer died)?

File: task-maker-async/src/store.rs
Comment (robryk) [inline task-maker-async/src/store.rs:90]: This should accept offset: otherwise it's not idempotent.

File: task-maker-async/src/store.rs
Comment (robryk) [inline task-maker-async/src/store.rs:44]: Present and finalized?

File: task-maker-async/src/store.rs
Comment (robryk) [inline task-maker-async/src/store.rs:25]: I don't understand what this enum is supposed to represent. Is OutputFile a full file (what is its name)? Is it the next chunk of output? What does Outcome mean? How does it interact with finalizing?

Maybe append_... is missing the argument with the actual data to append?

**Review by robryk — COMMENTED** [review]
First comments.

File: task-maker-async/src/store.rs
Comment (veluca93) [reply task-maker-async/src/store.rs:103]: We realize that when the lease expires, and at that point there shouldn't be another such computation, no?

File: task-maker-async/src/store.rs
Comment (veluca93) [reply task-maker-async/src/store.rs:25]: Ah, yes, indeed there is a missing argument, and OutputFile(String) is just the filename. Outcome means the "file" representing the outcome (memory / time used, return code,  ...)

File: task-maker-async/src/store.rs
Comment (veluca93) [reply task-maker-async/src/store.rs:44]: Right. What do you suggest we should do if present and *not* finalized? Return an error?

File: task-maker-async/src/store.rs
Comment (veluca93) [reply task-maker-async/src/store.rs:56]: Right. Modified to only refresh an input write lease, identified by a file_handle. (read_input_file gives you the option to get a read-only lease).

File: task-maker-async/src/store.rs
Comment (veluca93) [reply task-maker-async/src/store.rs:80]: Then attempts at reading it will return FileReadingOutcome::Dropped.  Makes sense?

File: task-maker-async/src/store.rs
Comment (veluca93) [reply task-maker-async/src/store.rs:44]: (this is how I modified the API for now)

File: task-maker-async/src/store.rs
Comment (robryk) [reply task-maker-async/src/store.rs:104]: Then reading from an in-progress computation gets very messy, because you don't know if you're reading files from the same attempt.

File: task-maker-async/src/store.rs
Comment (robryk) [reply task-maker-async/src/store.rs:81]: So, we'll read until the worker dies, and then we'll fail the next read, regardless of where in the file we were trying to read?

File: task-maker-async/src/store.rs
Comment (veluca93) [reply task-maker-async/src/store.rs:104]: How about dropping an in-progress computation invalidates all the handles to it, and this as a consequence makes all attempts to read those computations fail with an error?

File: task-maker-async/src/store.rs
Comment (veluca93) [reply task-maker-async/src/store.rs:81]: Yep -- see also other answer (note that until a computation is finalized, there is no EOF)

File: task-maker-async/src/store.rs
Comment (veluca93) [reply task-maker-async/src/store.rs:44]: Yes, I think that's fine. I'm thinking of leases being quite short (O(second)), so this shouldn't introduce noticeable delays.

File: task-maker-async/src/store.rs
Comment (robryk) [reply task-maker-async/src/store.rs:104]: Let's say that I want to read two files from the same computation. How can I guarantee that they come from the same attempt?

Comment (veluca93) [reply task-maker-async/src/store.rs:104]: Good question. How about getting a handle-to-computation which then can be used to access the files in the computation itself?

Comment (robryk) [reply task-maker-async/src/store.rs:104]: Yeah, and then we don't really need handles for individual files (though they might still be useful as a way to keep the read interface identical for both; that said, we should consider which handles holds leases and which take which others with them into explicit oblivion).

File: task-maker-async/src/store.rs
Comment (robryk) [inline task-maker-async/src/store.rs:54]: What prevents the file from disappearing immediately after it's written (after all, no one's holding a read lease)?

Similarly for computation results.

File: task-maker-async/src/store.rs
Comment (veluca93) [reply task-maker-async/src/store.rs:104]: I'd say that it's useful to keep per-file handles - but all the handles related to a computation share the same lease, and they all share the same fate

File: task-maker-async/src/store.rs
Comment (veluca93) [reply task-maker-async/src/store.rs:54]: Also a fair point. One option could be that finalizing an input file actually converts your write lease/handle into a read lease/handle.

The same wouldn't work for computations, but there the problem doesn't quite exist as the server (and the client) will be waiting for the computation to finish and obtain a read lease to it as soon as it's finalized.

Comment (veluca93) [reply task-maker-async/src/store.rs]: Made this change for input files.

File: task-maker-async/src/store.rs
Comment (veluca93) [reply task-maker-async/src/store.rs:113]: Clarified this.

**Comment by veluca93** [comment]
Also added a "kill computation" local method.

**Comment by veluca93** [comment]
One question I am unsure about: should there be only one lease type, with enum variant, and an error be returned if you use the wrong lease for an operation? This would simplify some of the interface, at the cost of introducing more errors. In particular it would allow to have a single "refresh" method.

Another option could be to have a Lease type, which every handle wraps, which can then be used for refreshes.

Thoughts?

File: task-maker-async/src/store.rs
Comment (edomora97) [inline task-maker-async/src/store.rs:15]: I get the point of having two hashes for this. I'm still not so sure on how `similar_computations` should be used though. It returns the set of `ComputationHash` sharing the first of the two hashes. But how can one get the information used for computing those hashes? I guess they are written somewhere as metadata file associated to a `ComputationHash`, but where?

**Review by edomora97 — COMMENTED** [review]
Plus a general note: very often you you used `()` as an error type. This is ok, but I kinda prefer using explicit error types (e.g. `anyhow::Error`) since they make debugging and tracing the error cause a lot easier.

**Comment by veluca93** [comment]
> Plus a general note: very often you you used `()` as an error type. This is ok, but I kinda prefer using explicit error types (e.g. `anyhow::Error`) since they make debugging and tracing the error cause a lot easier.

Yup, once this is actually implemented there will be a real error there ;)

File: task-maker-async/src/store.rs
Comment (veluca93) [reply task-maker-async/src/store.rs:61]: I don't think it can, changed

Comment (veluca93) [reply task-maker-async/src/store.rs:61]: (I swapped the replies to the two comments)

File: task-maker-async/src/store.rs
Comment (veluca93) [reply task-maker-async/src/store.rs:63]: Yup, leases are per-fileset

File: task-maker-async/src/server.rs
Comment (veluca93) [reply task-maker-async/src/server.rs:14]: Yes, using refresh_fileset_lease

File: task-maker-async/src/server.rs
Comment (veluca93) [reply task-maker-async/src/server.rs:23]: This API signals something by *answering* the request though :)

File: task-maker-async/src/server.rs
Comment (veluca93) [reply task-maker-async/src/server.rs:6]: Yep, sure - this is just a placeholder for now

File: task-maker-async/src/store.rs
Comment (veluca93) [reply task-maker-async/src/store.rs:5]: I'm not sure what you mean here - but I believe changing the array size should always be fine.

I used Vec<u8> because that's what we use today for file hashes ;)

File: task-maker-async/src/store.rs
Comment (veluca93) [reply task-maker-async/src/store.rs:15]: I added a second type of special files, other than MainFile, for `FileSet`s - a `Metadata` file type that will store this kind of information.

**Review by edomora97 — APPROVED** [review]

**Comment by edomora97** [comment]
I'll wait for the approval of @robryk before merging, in case there are still some comments to resolve.

(btw, thanks for looking into this!)

**Pull request merged**
